flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java

```diff
@@ -0,0 +1,563 @@
+
+                        return tryInjectRuntimeFilter(
+                                join.copy(join.getTraitSet(), Arrays.asList(newLeft, newRight)));
+                    }
+                };
+        return shuttle.visit(root);
+    }
+
+    /**
+     * Judge whether the join is suitable, and try to inject runtime filter for it.
+     *
+     * @param join the join node
+     * @return the new join node with runtime filter.
+     */
+    private static Join tryInjectRuntimeFilter(Join join) {
+
+        // check supported join type
+        if (join.getJoinType() != JoinRelType.INNER
+                && join.getJoinType() != JoinRelType.SEMI
+                && join.getJoinType() != JoinRelType.LEFT
+                && join.getJoinType() != JoinRelType.RIGHT) {
+            return join;
+        }
+
+        // check supported join implementation
+        if (!(join instanceof BatchPhysicalHashJoin)
+                && !(join instanceof BatchPhysicalSortMergeJoin)) {
+            return join;
+        }
+
+        boolean leftIsBuild;
+        if (canBeProbeSide(join.getLeft())) {
+            leftIsBuild = false;
+        } else if (canBeProbeSide(join.getRight())) {
+            leftIsBuild = true;
+        } else {
+            return join;
+        }
+
+        // check left join + left build
+        if (join.getJoinType() == JoinRelType.LEFT && !leftIsBuild) {
+            return join;
+        }
+
+        // check right join + right build
+        if (join.getJoinType() == JoinRelType.RIGHT && leftIsBuild) {
+            return join;
+        }
+
+        JoinInfo joinInfo = join.analyzeCondition();
+        RelNode buildSide;
+        RelNode probeSide;
+        ImmutableIntList buildIndices;
+        ImmutableIntList probeIndices;
+        if (leftIsBuild) {
+            buildSide = join.getLeft();
+            probeSide = join.getRight();
+            buildIndices = joinInfo.leftKeys;
+            probeIndices = joinInfo.rightKeys;
+        } else {
+            buildSide = join.getRight();
+            probeSide = join.getLeft();
+            buildIndices = joinInfo.rightKeys;
+            probeIndices = joinInfo.leftKeys;
+        }
+
+        Optional<BuildSideInfo> suitableBuildOpt =
+                findSuitableBuildSide(
+                        buildSide,
+                        buildIndices,
+                        (build, indices) ->
+                                isSuitableDataSize(build, probeSide, indices, probeIndices));
+
+        if (suitableBuildOpt.isPresent()) {
+            BuildSideInfo suitableBuildInfo = suitableBuildOpt.get();
+            RelNode newProbe =
+                    createNewProbeWithRuntimeFilter(
+                            ignoreExchange(suitableBuildInfo.buildSide),
+                            ignoreExchange(probeSide),
+                            suitableBuildInfo.buildIndices,
+                            probeIndices);
+            if (probeSide instanceof Exchange) {
```

Comment:
IMO, the current logic is that we inject the runtime filter node regardless of whether the probe side is an exchange node or not, can you give some comments to explain this?